File: 3Sum_4Sum/3sumClosest.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

int compare(const void* a, const void* b) {
  return (*(int*)a - *(int*)b);
}

int three_sum_closest(int* nums, int size, int target) {
  if (size < 3) {
    return 0;
  }

  qsort(nums, size, sizeof(int), compare);
  int closest = nums[0] + nums[1] + nums[2];

  for (int i = 0; i < size - 2; i++) {
    int second = i + 1, third = size - 1;

    while (second < third) {
      int sum = nums[i] + nums[second] + nums[third];
      if (sum == target) {
        return sum;
      }
      if (abs(target - sum) < abs(target - closest)) {
        closest = sum;
      } else if (sum < target) {
        second++;
      } else {
        third--;
      }
    }
  }
  return closest;
}

int main(int argc, char** argv) {
  int nums[] = {-1, 2, 1, -4};
  int size = sizeof(nums)/sizeof(nums[0]);
  int target = 1;

  printf("%d\n", three_sum_closest(nums, size, target));

  return 0;
}

File: 3Sum_4Sum/3sumClosest.rs
fn three_sum_closest(nums: &mut Vec<i32>, target: i32) -> i32 {
    if nums.len() < 3 {
        return 0;
    }

    nums.sort();

    let mut closest = nums[0] + nums[1] + nums[2];

    for i in 0..nums.len() - 2 {
        let mut second = i + 1;
        let mut third = nums.len() - 1;

        while second < third {
            let sum = nums[i] + nums[second] + nums[third];
            if sum == target {
                return sum;
            }
            if (target - sum).abs() < (target - closest).abs() {
                closest = sum;
            } else if sum < target {
                second += 1;
            } else {
                third -= 1;
            }
        }
    }
    closest
}

fn main() {
    let mut nums = vec![-1, 2, 1, -4];
    let target = 1;

    println!("{}", three_sum_closest(&mut nums, target));
}
